blocks/graph.py

```diff
@@ -161,8 +162,40 @@ def replace(self, replacements):
             substitutes.
 
         """
-        return ComputationGraph(theano.clone(self.outputs,
-                                             replace=replacements))
+        # Due to theano specifics we have to make one replacement in time
+        replacements = OrderedDict(replacements)
+
+        apply_nodes_sorted = theano.gof.graph.io_toposort(
+            self.inputs, self.outputs)
+        outputs_cur = self.outputs
+
+        # `replacements` with previous replacements applied. We have to track
+        # variables in the new graph corresponding to original replacements.
+        replacements_cur = OrderedDict()
+        # Sort `replacements` in topological order
+        for node in apply_nodes_sorted:
+            for input_ in node.inputs:
+                if input_ not in replacements:
+                    continue
+                replacements_cur[input_] = replacements[input_]
+
+        # Replace step-by-step in topological order
+        for i in range(len(replacements_cur)):
```

Comment:
This line and the following line could be replaced by the more idiomatic

```python
for i, replacement in enumerate(replacements_cur):
    ...
```
This should also be more efficient as it's not recreating the items list on every iteration.